### PR TITLE
Add bloom flax

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

Adds the Flax BLOOM model, superseding #18022 (where force pushing to Younes' branch after rebase closed the PR branch and created a new one)

### CODE REVIEW

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 1]
```diff
@@ -135,7 +135,19 @@ def is_key_or_prefix_key_in_dict(key: Tuple[str]) -> bool:
...
+    except ImportError:
+        logger.error(
+            "Loading a PyTorch model in Flax, requires both PyTorch and Flax to be installed. Please see"
+            " https://pytorch.org/ and https://flax.readthedocs.io/en/latest/installation.html for installation"
+            " instructions."
+        )
+        raise
+
+    is_bfloat_16 = all(v.dtype == torch.bfloat16 for v in pt_state_dict.values())  # noqa: F821
+    pt_state_dict = {k: v.numpy() if not is_bfloat_16 else v.float().numpy() for k, v in pt_state_dict.items()}
```
Comment (sgugger) [inline src/transformers/modeling_flax_pytorch_utils.py:150]: The test should be for each weight IMO as the state dict could contain weights in int8 (common for buffers) or even float32.
```suggestion
    pt_state_dict = {k: v.numpy() if not v.dtype == torch.bfloat16 else v.float().numpy() for k, v in pt_state_dict.items()}
```
and you can save a `weight_dtypes` dict that maps each weight name to its dtype.

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 2]
```diff
@@ -197,11 +209,15 @@ def convert_pytorch_state_dict_to_flax(pt_state_dict, flax_model):
                 continue
 
             # also add unexpected weight so that warning is thrown
-            flax_state_dict[("params",) + flax_key] = jnp.asarray(flax_tensor)
+            flax_state_dict[("params",) + flax_key] = (
+                jnp.asarray(flax_tensor) if not is_bfloat_16 else jnp.asarray(flax_tensor, dtype=jnp.bfloat16)
+            )
```
Comment (sgugger) [inline src/transformers/modeling_flax_pytorch_utils.py:214]: And then for here you can reuse the `weight_dtypes` dict to know if you need to convert back to bfloat16.

File: src/transformers/models/bloom/__init__.py [hunk 3]
```diff
@@ -14,11 +14,21 @@
...
+)
 
 
 _import_structure = {
-    "configuration_bloom": ["BLOOM_PRETRAINED_CONFIG_ARCHIVE_MAP", "BloomConfig", "BloomOnnxConfig"],
+    "configuration_bloom": [
+        "BLOOM_PRETRAINED_CONFIG_ARCHIVE_MAP",
+        "BloomConfig",
+        "BloomOnnxConfig",
+    ],
```
Comment (sgugger) [inline src/transformers/models/bloom/__init__.py:31]: Not sure why this change? It worked on one line.

File: tests/models/bloom/test_modeling_flax_bloom.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    from transformers import FlaxBloomForCausalLM, FlaxBloomModel
+
+
+def prepare_bloom_inputs_dict(config, input_ids, attention_mask=None):
+    if attention_mask is None:
+        attention_mask = np.where(input_ids != config.pad_token_id, 1, 0)
+    return {
+        "input_ids": input_ids,
+        "attention_mask": attention_mask,
+    }
```
Comment (sgugger) [inline tests/models/bloom/test_modeling_flax_bloom.py:44]: Nit: fits in one line.

File: src/transformers/models/bloom/modeling_flax_bloom.py [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+            # update key, value caches with our new 1d spatial slices
+            cur_index = cache_index.value
+            indices = (0,) * len(batch_dims) + (cur_index, 0, 0)
+            key = lax.dynamic_update_slice(cached_key.value, key, indices)
+            value = lax.dynamic_update_slice(cached_value.value, value, indices)
+            cached_key.value = key
+            cached_value.value = value
+            num_updated_cache_vectors = query.shape[1]
+            cache_index.value = cache_index.value + num_updated_cache_vectors
+            # causal mask for cached decoder self-attention: our single query position should only attend to those key positions that have already been generated and cached, not the remaining zero elements.
```
Comment (sgugger) [inline src/transformers/models/bloom/modeling_flax_bloom.py:210]: Can you put the comment on two lines to respect the 119 char limit?

**Review by sgugger — COMMENTED** [review]
Thanks for adding this new model in Flax! My main comment is in the changes in the modeling/Flax conversion.

File: src/transformers/models/bloom/__init__.py (hunk 3 above)
Comment (sanchit-gandhi) [reply src/transformers/models/bloom/__init__.py]: Interesting! It's back on one-line now: https://github.com/huggingface/transformers/pull/25094/commits/b6d8dc6b76ba1b979fea8045d4ab17e3d0f959c3

File: tests/models/bloom/test_modeling_flax_bloom.py (hunk 4 above)
Comment (sanchit-gandhi) [reply tests/models/bloom/test_modeling_flax_bloom.py]: Resolved in https://github.com/huggingface/transformers/pull/25094/commits/b6d8dc6b76ba1b979fea8045d4ab17e3d0f959c3

File: src/transformers/models/bloom/modeling_flax_bloom.py (hunk 5 above)
Comment (sanchit-gandhi) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: Resolved in https://github.com/huggingface/transformers/pull/25094/commits/b6d8dc6b76ba1b979fea8045d4ab17e3d0f959c3 by updating the Flax GPTJ modelling code (where we 'copy from')

**Review by sgugger — APPROVED** [review]
Thanks for iterating!

**Comment by sanchit-gandhi** [comment]
Thanks for the clinical review! Merging 👍
